Avoid Iteration on light travel time in Time peformance.rst

SG004 opened the pull request:
For Issue #7755

eteq commented on docs/time/performance.inc.rst:
```diff
@@ -10,3 +10,49 @@ Performance Tips
+>>> dec = np.random.normal(0.0, 1.0, 50000)
+>>> coos = coord.SkyCoord(ra, dec, unit=u.deg)
+
+   Lapalma
```
What is this supposed to be?  Is it meant to indicate that the location below is lapalma?  If so you should put it as either an inline comment or not be indented and be a full sentence.  Alternatively, the following might require *no* explanation:
```
observatory = coord.EarthLocation.of_site('lapalma')
```

SG004 replied:
`observatory = coord.EarthLocation.of_site('lapalma')`

Unfortunately, the code above creates an error in the circleCI test. That is why I chose to get the location the way I did.

eteq commented on docs/time/performance.inc.rst:
```diff
@@ -10,3 +10,49 @@ Performance Tips
+
+>>> observatory = coord.EarthLocation.from_geocentric(5327448.9957829, -1718665.73869569, 3051566.90295403, unit='m')
+
+*The first approach and its time to completion below:*
```
Why is this in italics?  Maybe just regular text is sufficient?  Or it could be a sub-sub-heading:
```
First approach
^^^^^^^^^^^^^^
```

SG004 replied:
I will turn it into a sub sub heading

eteq commented on docs/time/performance.inc.rst:
```diff
@@ -10,3 +10,49 @@ Performance Tips
+
+   **Wall time:** 58.2 ms
+
+*The second approach and its time to completion below:*
```
(same here as above)

eteq commented on docs/time/performance.inc.html:
```diff
@@ -0,0 +1,422 @@
+<?xml version="1.0" encoding="utf-8" ?>
```
I think you didn't mean to commit this file. The html is auto-generated by sphinx so it's not supposed to be in the source code repository.  If you can remove it from the git history that's best but if you're not sure how to do that, you can also just `git rm docs/time/performance.inc.html`  Just be careful not to commit it again!

SG004 replied:
I will remove it from being commited

astrofrog replied:
It looks like this was added back?

eteq commented on docs/time/performance.inc.rst:
```diff
@@ -10,3 +10,49 @@ Performance Tips
+As you can see, the user time is significantly longer. This is because iteration
+is being used.
+
+**AVOID ITERATION FOR LIGHT TRAVEL TIME FOR MANY OBJECTS**
```
Might also be best as a sub-sub-section?

eteq commented on docs/time/performance.inc.rst:
```diff
@@ -10,3 +10,49 @@ Performance Tips
 
 Here we provide some tips and tricks for how to optimize performance of code
 using `astropy.time`.
+
```
You should have an rst-style sub-header here.  Something like

```
Light Travel Times
------------------
```

StuartLittlefair commented on docs/time/performance.inc.rst:
```diff
@@ -10,3 +10,52 @@ Performance Tips
+>>> ra = np.random.normal(0.0, 1.0, 50000)
+>>> dec = np.random.normal(0.0, 1.0, 50000)
+>>> coos = coord.SkyCoord(ra, dec, unit=u.deg)
+>>> observatory = coord.EarthLocation.from_geocentric(5327448.9957829, -1718665.73869569, 3051566.90295403, unit='m') #Lapalma Observatory
```
Perhaps you can put the end-of-line comment above in it's own line, and perhaps note that 
```coord.EarthLocation.of_site('lapalma')``` would also work if you have internet connection

SG004 replied:
I will do that!

StuartLittlefair commented on docs/time/performance.inc.rst:
```diff
@@ -10,3 +10,52 @@ Performance Tips
+
+   **Wall time:** 16min 58s
+
+As you can see, the user time is significantly longer.
```
Perhaps explain why? Iterating over a ```SkyCoord``` is not fast, and the former approach also takes advantage of vectorised operations in numpy...

SG004 replied:
Sure, no problem!

StuartLittlefair commented on docs/time/performance.inc.rst:
```diff
@@ -10,3 +10,52 @@ Performance Tips
+For user cases where there are thousands of times for each source, broadcasting
+can be used:
+
+>>> times = Time.now() + np.linspace(0, 3, 1000)*u.day
```
I think we probably need to spell out this example quite a bit more. At the moment it's not quite clear what's going on to anyone without a detailed knowledge of numpy's broadcasting rules. 

A minimal change would be to separate out the step adding a new axis to ```coos```, show the shape afterwards and explain that the change of shape means that the 1000 times and the 50,000 co-ordinates will be broadcast against each other. You could possibly link to external information on Numpy broadcasting. 

At the end, I would show the shape of the resulting ```ltts``` object, since it will help clarify what has happened.

SG004 commented on docs/time/performance.inc.rst:
```diff
@@ -10,3 +10,49 @@ Performance Tips
 Here we provide some tips and tricks for how to optimize performance of code
 using `astropy.time`.
+
+Listed below are two approaches to calculating light travel times for tens of
```
Will do!

SG004 commented:
Hey! I've updated with most of the requested changes, however there are a few I could not do. I have run into a problem with the light_travel_time function. Everywhere in the code that calls it, I have errors when running the doctests. I will add a couple screenshots of the errors I keep getting.
The first call to light_travel_time on line 41. I used explicitly passed in the skycoord argument.
![error when explictly passing arguments](https://user-images.githubusercontent.com/28075025/47758584-00ad0b80-dc82-11e8-940b-cd21cd666e86.PNG)

The second call using the iteration method on line 51. I did not explicitly pass the skycoord argument
![not explicitly calling arguments](https://user-images.githubusercontent.com/28075025/47758667-63060c00-dc82-11e8-8090-b19b3bc78e86.PNG)

kakirastern commented:
Hi, just am performing a bit of triaging... I see some inactivity in this PR. @SG004, would you like me to try and reproduce the problem to help troubleshoot, or even take over to free you to other tasks if you prefer? Please let me know... (I might need to open another PR to take over if that is the case.)

SG004 commented:
> Hi, just am performing a bit of triaging... I see some inactivity in this PR. @SG004, would you like me to try and reproduce the problem to help troubleshoot, or even take over to free you to other tasks if you prefer? Please let me know... (I might need to open another PR to take over if that is the case.)

I am all for you taking over for this, I wouldn't know what exactly to do to "hand it over", but yeah you are welcome to it.

kakirastern commented:
No worries, I will simply need to open another PR and start everything from scratch based on what you have done, with some modifications where needed. I will definitely let you know if I will stumble on the same errors or if I do find out what went wrong with your commits however...

kakirastern commented:
Hi @SG004! I think there have been some bugs in the original code snippets provided, and one of the positional arguments have been omitted, if you try the following the code might work... 

Code block 1:
```
>>> import numpy as np
>>> import astropy.coordinates as coord
>>> import astropy.units as u
>>> from astropy.time import Time

>>> ra = np.random.normal(0.0, 1.0, 50000)
>>> dec = np.random.normal(0.0, 1.0, 50000)

>>> coos = coord.SkyCoord(ra, dec, unit=u.deg)
>>> observatory = coord.EarthLocation.of_site('lapalma')

>>> %time Time.light_travel_time(Time.now(), coos, location=observatory)

>>> %time ltts = [Time.light_travel_time(Time.now(), coo, location=observatory) for coo in coos]
``` 

Code block 2 as is should be fine, just as a recap as follows:
```
>>> times = Time.now() + np.linspace(0, 3, 1000)*u.day
>>> %time ltts = times.light_travel_time(coos[:, np.newaxis], location=observatory)
```

kakirastern commented:
So if you are still interested you can keep up the good work and continue until merge. Let me know what you think. Cheers!

SG004 commented:
> So if you are still interested you can keep up the good work and continue until merge. Let me know what you think. Cheers!

So, I realized I deleted the repository that I had forked back in October and the branch for this PR was deleted, so I will have to start a new PR. That being said I copied and pasted my old code, and swapped out the code blocks for yours, but when I ran a doctest, it still fails.

https://gist.github.com/SG004/9a8904c15ed581c45e8f6cba0ba0c532

kakirastern commented:
Please try the following: 
Error 1: Get rid of `%time` and run again
Error 2: Similar, get rid of `%time` and run again
Error 3: Add self=Time.now() to first positional argument of light_travel_time() method
Error 4: Similar, add self=Time.now() to first positional argument of light_travel_time() method
Error 5: Trying running both of the following at the same command prompt:
```
times = Time.now() + np.linspace(0, 3, 1000)*u.day
ltts = times.light_travel_time(coos[:, np.newaxis], location=observatory)
```
Don't separate out using `coos2` yet... If successful, proceed to further debugging. 
BTW, I think `coos2 = coos[:, np.newaxis]` and should be sandwiched in between the `times` definition and `ltts` lines. I am not sure hence I am stating it here...

kakirastern commented:
If need be you might want to check the astropy version you have installed, and change to one that works like the stable v3.1.1 version. It's important to check since I suspect there has been some error importing `from astropy.time import Time` for you... Or, it could simply be due to typo(s), so be careful with those.

kakirastern commented:
And it might be advisable to use Python 3.6 at least for now... if 3.7 if not working well for you.

SG004 commented:
> And it might be advisable to use Python 3.6 at least for now... if 3.7 if not working well for you.

So, I've switched to a python 3.6 environment and I've addressed your suggestions. Linked below is the current file and the current failing test output for it.

https://gist.github.com/SG004/b20aa10f247f081c0fda3a22cbda2333

kakirastern commented:
So one quick bug I spotted is `Time.light_travel_time(self=Time.now,skycoord=coos, location=observatory )` in L31, as there are missing round brackets in `Time.now()'.

kakirastern commented:
Also, try getting rid of the space after the positional argument `location=observatory` in `Time.light_travel_time(self=Time.now,skycoord=coos, location=observatory )`... (not sure if that's why there is an error associated with it though, but am sure no harm will be done if the space is not there).

kakirastern commented:
The other three are all MemoryError's. It could be due to the setup of your computer. Have you tried running the code on Google's Colaboratory before and see if you could duplicate the error. You might need a Google account if you don't already have one. It can be accessed at [https://colab.research.google.com/](https://colab.research.google.com/). I am not sure if you are using the 64-bit version of Python 3.6... unless the issue is that your computer ran out of RAM. Try it and see if it works there.

kakirastern commented:
And, I really would recommend using a virtual environment created with something like conda to help you to manage your system setup. More info on that can be found at [http://docs.astropy.org/en/stable/development/workflow/virtual_pythons.html#set-up-for-virtual-environments](http://docs.astropy.org/en/stable/development/workflow/virtual_pythons.html#set-up-for-virtual-environments). I am currently using it for my work with the development version of astropy, because it simplifies the task of managing your software installations as it checks for any dependencies and their compatibility for you. 

If you are indeed using a virtual environment now then there might be something wrong with the Python version you are using. Generally check the executable by typing `which python` and `which python3` (note they might be different), and the version by either `python --version` or `python3 --version`, depending on your system configuration.

kakirastern commented:
And, it might be a good idea to add back `%time` to some of the lines if possible, granted that now that the code works.

SG004 commented:
> And, it might be a good idea to add back `%time` to some of the lines if possible, granted that now that the code works.

@kakirastern Thanks for the recommendation of Google's colaboratory. I fixed the typos and then ran the code in the colab. Unfortunately, even with the colab's 12GB of RAM, the iteration over the skycoord ended up creating a memory error. I lowered the number of iterations to 20,000 from 50,000 and colab was able to run it in approximately 12 minutes. And I spent the evening looking into virtual environments and have setup up a couple conda environments. I will submit a new PR with all of the changes and the working code sometime tomorrow.

kakirastern commented:
@SG004 That's good to know you have figured out how to solve the problem. Remember to let the admins know you have a new PR open so they will know to close this one. Cheers! 

(But just check your remote to see if the fork is still there, which I suspect it is... This is because if the fork is gone from your remote it should show up in this PR, which has not yet happened yet. Check that with the command `git remote -v` at the terminal once you are in your Git astropy folder. If your branch is still here, you will simply need to do `git branch (name of your branch)` and then go `git checkout (name of your branch)`, and follow the rest of the steps [here](http://www.gemini.edu/sciops/observing-gemini/2018b-call-proposals) to recover it. That way you will not need to open another PR.)

SG004 commented:
> @SG004 That's good to know you have figured out how to solve the problem. Remember to let the admins know you have a new PR opened so they will know to close this one. Cheers!
> 
> (But just check your remote to see if the fork is still there, which I suspect it is... This is because if the fork is gone from your remote it should show up in this PR, which has not yet happened yet. Check that with the command `git remote -v` at the terminal once you are in your Git astropy folder. If your branch is still here, you will simply need to do `git branch (name of your branch)` and then go `git checkout (name of your branch)`, and follow the rest of the steps [here](http://www.gemini.edu/sciops/observing-gemini/2018b-call-proposals) to recover it. That way you will not need to open another PR.)

I deleted my fork and re-forked recently, so my branch does not exist anymore. It looks like I will have to open another PR

P.S Sorry for taking so long to reply, I caught a stomach bug and haven't been near my computer the past couple days

pllim commented:
OK I am closing this then, given it is impossible to update this fork/branch anymore. Please do refer to this PR number when you get around to submitting a new one. Hope you feel better soon!

kakirastern commented:
@SG004 Get well soon!
